components/director/internal/tenantfetchersvc/subscriber.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/pkg/errors"
+)
+
+// TenantProvisioner is used to create all related to the incoming request tenants, and build their hierarchy;
+//go:generate mockery --name=TenantProvisioner --output=automock --outpkg=automock --case=underscore
+type TenantProvisioner interface {
+	ProvisionTenants(context.Context, *TenantSubscriptionRequest, string) error
+}
+
+// RuntimeService is used to interact with runtimes
+//go:generate mockery --name=RuntimeService --output=automock --outpkg=automock --case=underscore
+type RuntimeService interface {
+	SetLabel(context.Context, *model.LabelInput) error
+	GetLabel(ctx context.Context, runtimeID string, key string) (*model.Label, error)
+	ListByFiltersGlobal(context.Context, []*labelfilter.LabelFilter) ([]*model.Runtime, error)
+}
+
+type subscriptionFunc func(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error
+
+type sliceMutationFunc func([]string, string) []string
+
+type subscriber struct {
+	provisioner TenantProvisioner
+	RuntimeService
+	RegionLabelKey                string
+	SubscriptionConsumerLabelKey  string
+	ConsumerSubaccountIDsLabelKey string
+}
+
+// NewSubscriber creates new subscriber
+func NewSubscriber(provisioner TenantProvisioner, service RuntimeService, regionLabelKey, subscriptionConsumerLabelKey, consumerSubaccountIDsLabelKey string) *subscriber {
+	return &subscriber{
+		provisioner:                   provisioner,
+		RuntimeService:                service,
+		RegionLabelKey:                regionLabelKey,
+		SubscriptionConsumerLabelKey:  subscriptionConsumerLabelKey,
+		ConsumerSubaccountIDsLabelKey: consumerSubaccountIDsLabelKey,
+	}
+}
+
+// Subscribe subscribes tenant to runtime. If the tenant does not exist it will be created
+func (s *subscriber) Subscribe(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error {
+	if err := s.provisioner.ProvisionTenants(ctx, tenantSubscriptionRequest, region); err != nil {
+		return err
+	}
+
+	return s.applyRuntimesSubscriptionChange(ctx, tenantSubscriptionRequest.SubscriptionConsumerID, tenantSubscriptionRequest.SubaccountTenantID, region, addElement)
+}
+
+// Unsubscribe unsubscribes tenant from runtime.
+func (s *subscriber) Unsubscribe(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error {
+	return s.applyRuntimesSubscriptionChange(ctx, tenantSubscriptionRequest.SubscriptionConsumerID, tenantSubscriptionRequest.SubaccountTenantID, region, removeElement)
+}
+
+func (s *subscriber) applyRuntimesSubscriptionChange(ctx context.Context, subscriptionConsumerID, subaccountTenantID, region string, mutateSliceFunc sliceMutationFunc) error {
+	filters := []*labelfilter.LabelFilter{
+		labelfilter.NewForKeyWithQuery(s.SubscriptionConsumerLabelKey, fmt.Sprintf("\"%s\"", subscriptionConsumerID)),
```

Comment:
Shouldn't `subscriptionConsumerID` be `subscriptionProviderId` as the provider is the one having the runtime and the consumer is the one consuming the runtime? 
Perhaps the naming of the properties should be aligned somehow.